src/library/vm/vm_cached.cpp

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Mario Carneiro
+*/
+#include <vector>
+#include "library/vm/vm.h"
+#include "library/trace.h"
+
+#define lean_cached_trace(CODE) lean_trace(name({"cached", "update"}), CODE)
+
+namespace lean {
+struct vm_cached : public vm_external {
+    vm_obj m_default;
+    vm_obj m_value;
+    vm_cached(vm_obj const & def, vm_obj const & v):m_default(def), m_value(v) {}
+    virtual ~vm_cached() {}
+    virtual void dealloc() override { this->~vm_cached(); get_vm_allocator().deallocate(sizeof(vm_cached), this); }
+    virtual vm_external * ts_clone(vm_clone_fn const &) override { return new vm_cached(m_default, m_value); }
```

Comment:
`m_default` is the passed in `inhabited A` instance, which may possibly be used in the future if we implement resetting the cache during garbage collection. Note that even if the class is ditched, the pointer indirection is essential for making this work, although a bare pointer could probably do the job just as well.